main/src/ca/uwaterloo/flix/language/phase/unification/set/SetUnification.scala

```diff
@@ -0,0 +1,235 @@
+        case (ElemSet(i1), ElemSet(i2)) if i1 != i2 => error()
+        case (ElemSet(_), Cst(_)) => error()
+        case (Cst(_), Univ) => error()
+        case (Cst(_), Empty) => error()
+        case (Cst(_), ElemSet(_)) => error()
+        case (Cst(c1), Cst(c2)) if c1 != c2 => error()
+        // Equations that are not trivial.
+        case _ => (List(eq), SetSubstitution.empty) // Can't do anything.
+      }
+    }
+
+    /**
+      * Solves equations of ground assignments to variables (e.g. `x ~ c1 ∪ e2`).
+      *
+      *   - `x ~ t` where [[SetFormula.isGround]] on `t` becomes `({}, [x -> t])`
+      *   - `!x ~ t` where [[SetFormula.isGround]] on `t` becomes `({}, [x -> !t])`
+      *   - `t1 ∩ t2 ∩ .. ~ univ` becomes `({t1 ~ univ, t2 ~ univ, ..}, [])`
+      *   - `t1 ∪ t2 ∪ .. ~ univ` becomes `({t1 ~ empty, t2 ~ empty, ..}, [])`
+      */
+    def constantAssignment(eq: Equation)(implicit p: Progress): (List[Equation], SetSubstitution) = {
+      val Equation(f1, f2, _, loc) = eq
+      (f1, f2) match {
+        // x ~ t, where t is ground
+        // ---
+        // {},
+        // [x -> t]
+        case (Var(x), f) if f.isGround =>
+          p.progress()
+          (Nil, SetSubstitution.singleton(x, f))
+
+        // !x ~ t, where t is ground
+        // ---
+        // {},
+        // [x -> !t]
+        case (Compl(Var(x)), f) if f.isGround =>
+          p.progress()
+          (Nil, SetSubstitution.singleton(x, mkCompl(f)))
+
+        // f1 ∩ f2 ∩ .. ~ univ
+        // ---
+        // {f1 ~ univ, f2 ~ univ, ..},
+        // []
+        case (inter@Inter(_, _, _, _, _, _, _), Univ) =>
+          p.progress()
+          val eqs = inter.mapSubformulas(Equation.mk(_, Univ, loc))
+          (eqs, SetSubstitution.empty)
+
+        // f1 ∪ f2 ∪ ... ~ empty
+        // ---
+        // [f1 ~ empty, f2 ~ empty, ...],
+        // {}
+        case (union@Union(_, _, _, _, _, _, _), Empty) =>
+          p.progress()
+          val eqs = union.mapSubformulas(Equation.mk(_, Empty, loc))
+          (eqs, SetSubstitution.empty)
+
+        // f1 ~ f2, where f1 and f2 are ground
+        // ---
+        // [], {} if solved
+        // [f1 !~ f2], {} if unsolvable
+        case (f1, f2) if f1.isGround && f2.isGround =>
+          p.progress()
+          if (isEquivalent(f1, f2)) (Nil, SetSubstitution.empty)
+          else (List(eq.toUnsolvable), SetSubstitution.empty)
+
+        case _ =>
+          // Can't do anything.
+          (List(eq), SetSubstitution.empty)
+      }
+    }
+
+    /**
+      * Solves equations variable alias equations (e.g. `x1 ~ x2`).
+      *
+      *   - `x1 ~ x2` becomes `({}, [x1 -> x2])`
+      *   - `!x1 ~ !x2` becomes `({}, [x1 -> x2])`
+      */
+    def variableAlias(eq: Equation)(implicit p: Progress): (List[Equation], SetSubstitution) = {
+      val Equation(t1, t2, _, _) = eq
+      (t1, t2) match {
+        // x1 ~ x2
+        // ---
+        // {},
+        // [x1 -> x2]
+        case (Var(x), y@Var(_)) =>
```

Comment:
What about literal x -> x? Is that gone? Can it happen here?

Reply:
if you mean `x42 ~ x42` then that is taken care of by `trivial` but we could also add it here yes

Reply:
OK, I leave it to you.